### PR TITLE
KV: make sure iterators aren't opened to be closed

### LINKED ISSUE #3884: KV Iterators are created twice
When SeekGE is being used after creating the iterator, the first creation of the iterator is meaningless and adds an overhead of a KV list call

### PR DESCRIPTION
**Pull request by itaiad200**
closes #3884

### CODE REVIEW

File: pkg/kv/iterators.go [hunk 1]
```diff
@@ -253,24 +253,29 @@ type PartitionIterator struct {
 }
 
 func NewPartitionIterator(ctx context.Context, store Store, msgType protoreflect.MessageType, partitionKey string) (*PartitionIterator, error) {
```
Comment (N-o-Z) [inline pkg/kv/iterators.go:255]: In that case - so why do we need to return an error?

File: pkg/kv/iterators.go [hunk 2]
```diff
@@ -253,24 +253,29 @@ type PartitionIterator struct {
...
 	return &PartitionIterator{
 		ctx:          ctx,
 		store:        store,
 		msgType:      msgType,
-		itr:          itr,
 		partitionKey: partitionKey,
 	}, nil
 }
 
+func getNewIter(ctx context.Context, store Store, partitionKey string, start []byte) (EntriesIterator, error) {
```
Comment (N-o-Z) [inline pkg/kv/iterators.go:264]: getNewEntriesIter?

**Review by N-o-Z — CHANGES_REQUESTED** [review]
Few comments

File: pkg/catalog/catalog.go [hunk 3]
```diff
@@ -764,7 +764,10 @@ func (c *Catalog) ListEntries(ctx context.Context, repository string, reference
 	it := NewEntryListingIterator(NewValueToEntryIterator(iter), prefixPath, delimiterPath)
 	defer it.Close()
 
-	it.SeekGE(afterPath)
+	if afterPath != "" {
+		it.SeekGE(afterPath)
+	}
```
Comment (nopcoder) [inline pkg/catalog/catalog.go:769]: out of iterator optimization :)

File: pkg/catalog/prefix_iterator.go [hunk 4]
```diff
@@ -13,7 +13,9 @@ type prefixIterator struct {
 }
 
 func NewPrefixIterator(it EntryIterator, prefix Path) EntryIterator {
-	it.SeekGE(prefix)
+	if len(prefix) > 0 {
+		it.SeekGE(prefix)
+	}
```
Comment (nopcoder) [inline pkg/catalog/prefix_iterator.go:18]: Think this one can cause a bug - we assume we accept a fresh 'it' and we do not need to call SeekGE, but we must

File: pkg/kv/iterators.go (hunk 2 above)
Comment (nopcoder) [inline pkg/kv/iterators.go:264]: Not sure we need a function that calls a function - if it was a member to capture the use of: p.{ctx, store, paritionKey} it was a different story.

File: pkg/kv/iterators.go [hunk 5]
```diff
@@ -253,24 +253,29 @@ type PartitionIterator struct {
...
 	if p.Err() != nil {
 		return false
 	}
 	p.value = nil
+	if p.itr == nil {
+		p.itr, p.err = getNewIter(p.ctx, p.store, p.partitionKey, []byte(""))
+		if p.Err() != nil {
+			return false
+		}
+	}
```
Comment (nopcoder) [inline pkg/kv/iterators.go:278]: In the case 'p.itr' is nil it means we failed to initialize or seek to a key - we should render an error or return false.

**Review by nopcoder — CHANGES_REQUESTED** [review]

File: pkg/catalog/prefix_iterator.go (hunk 4 above)
Comment (itaiad200) [reply pkg/catalog/prefix_iterator.go]: Agreed, fixed on the caller side.

File: pkg/kv/iterators.go [hunk 6]
```diff
@@ -252,25 +252,26 @@ type PartitionIterator struct {
 	err          error
 }
 
-func NewPartitionIterator(ctx context.Context, store Store, msgType protoreflect.MessageType, partitionKey string) (*PartitionIterator, error) {
-	itr, err := store.Scan(ctx, []byte(partitionKey), []byte(""))
-	if err != nil {
-		return nil, err
-	}
+func NewPartitionIterator(ctx context.Context, store Store, msgType protoreflect.MessageType, partitionKey string) *PartitionIterator {
```
Comment (N-o-Z) [inline pkg/kv/iterators.go:255]: I think the issue with SeekGE is relevant also for KVOrderedCommitIterator

**Review by N-o-Z — APPROVED** [review]
See additional comment

File: pkg/kv/iterators.go (hunk 6 above)
Comment (itaiad200) [reply pkg/kv/iterators.go:255]: That's a GC iterator, right? I don't assume it happens that often

**Pull request merged**